Issue #3134: Add AnnotationDefHandler for custom annotations

kukreja-vikramaditya opened the pull request:
Issue #3134

kukreja-vikramaditya commented:
I am facing some issue with code coverage. The `PrimordialHandler`'s **getIndentImpl()** method is not being covered. I don't understand the reason as I did not make any changes to that? Am I missing some test cases? Should I copy my Input file to the `InputDifficultAnnotations.java` file instead and use that?

rnveach commented:
@kukreja-vikramaditya Since coverage is a missing method, if you reset back to master run all the tests and set a breakpoint on the method you can see which test(s) was calling the method.

`testAnnotationDefinition` is what hits that method. Going back through the stack trace, it is called from `ObjectBlockHandler` for the AST at line 3 column 32 inside `InputAnnotationDefinition` which is an annotation interface.

rnveach commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/InputCustomAnnotation.java:
```diff
@@ -0,0 +1,83 @@
+String author() default "author"; //indent:0 exp:4 warn
+     } //indent:5 exp:0 warn
+
+     @interface //WRONG //indent:5 exp:5
```
why are we accepting this if it is wrong?

kukreja-vikramaditya replied:
The problem is arising as there is no dedicated handler for the` ANNOTATION_DEF` token. The `ClassDefHandler` doesn't check for `@interface` token. Do you think this should be a separate issue if this PR is merged? Or I should fix it with this PR as after all it is a part of the problem?

rnveach replied:
@kukreja-vikramaditya Since `ClassDefHandler` is used by other tokens, not just annotations, I am fine with the fix being inside `ClassDefHandler` and not having a separate handler.
Please see if you can fix this as part of this PR.

kukreja-vikramaditya replied:
Yes I think it would best as I only need to modify a few lines and check for `@interface`

rnveach commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/InputCustomAnnotation.java:
```diff
@@ -0,0 +1,83 @@
+   @Retention(RetentionPolicy.SOURCE) //indent:3 exp:4 warn
+@Repeatable(RepeatableInner.class) //indent:0 exp:4 warn
+@interface //indent:0 exp:0
+        AnnotationInnerLineWrap //WRONG exp 8 //indent:8 exp:7 warn
```
We don't need these `WRONG` comments if there is a warning/violation on the line.

rnveach commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/InputCustomAnnotation.java:
```diff
@@ -0,0 +1,83 @@
+
+     interface TestInterface {} //indent:5 exp:0 warn
+
+     enum TestEnumInput {} //indent:5 exp:0 warn
```
We can remove anything that isn't helping an annotation with it's test.
I just added these in issue to showcase we get violations other block types correctly.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/AnnotationDefHandler.java:
```diff
@@ -0,0 +1,91 @@
+     */
+    public AnnotationDefHandler(IndentationCheck indentCheck,
+                            DetailAST ast, AbstractExpressionHandler parent) {
+        super(indentCheck, "annotation def", ast, parent);
```
as noted below/before, this is the field defintition, not the annotation definition.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/HandlerFactory.java:
```diff
@@ -74,6 +74,8 @@ public HandlerFactory() {
         register(TokenTypes.LITERAL_SYNCHRONIZED, SynchronizedHandler.class);
         register(TokenTypes.LAMBDA, LambdaHandler.class);
+        register(TokenTypes.ANNOTATION_DEF, ClassDefHandler.class);
+        register(TokenTypes.ANNOTATION_FIELD_DEF, AnnotationDefHandler.class);
```
Separate question. For annotation definition you use `ClassDefHandler`.
`AnnotationDefHandler` looks like it coded like `MemberDefHandler`.
We can't use that handler, or give them common parents, instead? Does it make sense to do this?

kukreja-vikramaditya replied:
>AnnotationDefHandler looks like it is coded like MemberDefHandler.

Yes I based the field checking on member def handler.

The only case that the current `ClassDefHandler` is not solving for Annotation definition is line wrapping as you pointed out it in https://github.com/checkstyle/checkstyle/pull/4217#discussion_r111635354

kukreja-vikramaditya commented:
@rnveach 
Reports at https://kukreja-vikramaditya.github.io/checkstyle-diff-reports/fix-issue-3134/diff/index.html There are cases of annotation def where the `ClassDefHandler` cannot handle it's wrapping. This may be related to comment at https://github.com/checkstyle/checkstyle/pull/4217#discussion_r111635354 I'll check the reports soon and share the results

How should i proceed regarding this? Fix it in the `ClassDefHandler` directly?

rnveach commented:
> How should i proceed regarding this?

See my comment at https://github.com/checkstyle/checkstyle/pull/4217#discussion_r111683471 .

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/HandlerFactory.java:
```diff
@@ -74,6 +74,8 @@ public HandlerFactory() {
         register(TokenTypes.LITERAL_SYNCHRONIZED, SynchronizedHandler.class);
         register(TokenTypes.LAMBDA, LambdaHandler.class);
+        register(TokenTypes.ANNOTATION_DEF, ClassDefHandler.class);
+        register(TokenTypes.ANNOTATION_FIELD_DEF, AnnotationFieldDefHandler.class);
```
I still didn't receive an answer or change from my previous comment. https://github.com/checkstyle/checkstyle/pull/4217#discussion_r111636791

> > AnnotationDefHandler looks like it is coded like MemberDefHandler.
We can't use that handler, or give them common parents, instead? Does it make sense to do this?

> Yes I based the field checking on member def handler.

That makes it sound more like they should have comment class hierarchy.

kukreja-vikramaditya replied:
@rnveach 
It would actually be possible to get rid of `AnnotationFieldDefHandler`  completely and simply just edit the `checkIndentation` method of `MemberDefHandler` (at [L69](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/MemberDefHandler.java#L69) and [L56](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/MemberDefHandler.java#L56) ), unless future additions/changes are required.

I will share my results on the reports soon. Hopefully a better decision can be taken based on the report

rnveach replied:
http://checkstyle.sourceforge.net/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#ANNOTATION_FIELD_DEF
Since they are more methods than variables, should we base it on `MethodDefHandler` instead?

kukreja-vikramaditya commented:
@rnveach
I have updated the PR and reports [here](https://kukreja-vikramaditya.github.io/checkstyle-diff-reports/fix-issue-3134/diff/index.html) .

All except `findbugs` and `spring framework` are correctly showing errors.
The issue in both is same. Class inside interface definition.

How should this be handled? I think I am missing some `checkIndentaion` inside the OBJBLOCK in `ClassDefHandler`, but wouldn't that interfere with the checking by  `AnnotationFieldDefHandler` when annotation fields are present inside object block?

rnveach commented:
Sorry for delay.

> All except findbugs and spring framework are correctly showing errors.
> The issue in both is same. Class inside interface definition.
> How should this be handled? 

Can you please give explicit links to the violation and link to line in code for the problem you are talking about, and why indentation value is wrong and what you think it should be. It is not clear which one you think is a problem of the group.

I assume you are talking about these groups:
https://kukreja-vikramaditya.github.io/checkstyle-diff-reports/fix-issue-3134/diff/findbugs/index.html#A14
https://kukreja-vikramaditya.github.io/checkstyle-diff-reports/fix-issue-3134/diff/findbugs/xref/home/vikramaditya/Desktop/CS/contribution-master/checkstyle-tester/repositories/findbugs/findbugsTestCases/src/java/jsr305/validation/CheckValidatorSandboxing.java.html#L46
https://kukreja-vikramaditya.github.io/checkstyle-diff-reports/fix-issue-3134/diff/spring-framework/index.html#A1
https://kukreja-vikramaditya.github.io/checkstyle-diff-reports/fix-issue-3134/diff/findbugs/xref/home/vikramaditya/Desktop/CS/contribution-master/checkstyle-tester/repositories/findbugs/findbugsTestCases/src/java/jsr305/validation/CheckValidatorSandboxing.java.html#L46

but I am not sure yet which one you think is wrong.

This is `class` inside `@interface` with another `class` inside it. Just eye-balling the code, it looks all formatted correctly. So I wouldn't expect violations.
Each class/`@interface` level increases the indentation by 4 spaces or 1 tab. Base class should be at 0, nested `@interface` should be at 4, nested class should be at 8. Members of the class/`@interface` should be at 4 + the class/`@interface` requested indentation.
The reason these violations go away is because without a handler for `@interface`s, it doesn't see the brace and therefore doesn't increment requested indentation. But you added the handler now, so it is seeing that it needs to increase indentation for each nesting.

Let me know if I am missing something.

kukreja-vikramaditya commented:
@rnveach
I checked `findbugs` and it agrees with your review.

For `spring framework`, there are two classes defintions and 1 array declarator field.
Now in my PR I know that array declarator cases are not handled. But what about the class definitions? Shouldn't they be +4 from the interface definition?

https://kukreja-vikramaditya.github.io/checkstyle-diff-reports/fix-issue-3134/diff/spring-framework/xref/home/vikramaditya/Desktop/CS/contribution-master/checkstyle-tester/repositories/spring-framework/spring-test/src/test/java/org/springframework/test/util/MetaAnnotationUtilsTests.java.html#L453

rnveach commented:
> But what about the class definitions? Shouldn't they be +4 from the interface definition?

> >  nested @interface should be at 4, nested class should be at 8.

Yes, this is what I was referring to by "nested class", the class inside `@interface` inside `class`.
That code looks correctly indented to me. It is using tabs, but 1 tab is the same as 4 spaces (defined by your `tabWidth` in configuration) and spring uses 1 tab there (which I can tell by highlighting the whitespace).

kukreja-vikramaditya commented:
@rnveach
My bad.
Please do final review if any things need to be edited.
Thanks!

rnveach commented:
@kukreja-vikramaditya So far I think the reports look good.

I haven't see anything for the change we talked about at https://github.com/checkstyle/checkstyle/pull/4217#discussion_r111683553 .

kukreja-vikramaditya commented:
>Since they are more methods than variables, should we base it on MethodDefHandler instead?

Did you mean `MemberDefHandler`?

>I haven't see anything for the change we talked about at #4217 (comment) .

Yes I think that would be the best route to take since most of the methods are same with a few minor changes. I'll make it extend `MemberDefHandler` and generate reports to confirm there is no regression.

rnveach commented:
> > Since they are more methods than variables, should we base it on MethodDefHandler instead?

> Did you mean MemberDefHandler?

Look at `HandlerFactory` and http://checkstyle.sourceforge.net/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html.
`MethodDefHandler` is connected to `METHOD_DEF` which is for methods.
`MemberDefHandler` is connected to `VARIABLE_DEF` which is for local variables and class/enum/etc fields.

Going by http://checkstyle.sourceforge.net/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#ANNOTATION_FIELD_DEF , `ANNOTATION_FIELD_DEF` more closely resembles a `METHOD_DEF` instead of a `VARIABLE_DEF`, right?

kukreja-vikramaditya commented:
@rnveach 
I see the reason for basing it on `MethodDefHandler`. I extended it to MethodDefHandler and `checkModifiers()` was removed without any issue. I'll try to do the same for `checkType()` also, at the most a few changes would be required in MethodDefHandler.

I have a ques regarding the indentation of line 4,
```java
@interface HelloAnnot { //indent:0 exp:0
    public //indent:4 exp:4
        String myString() //indent:8 exp:8
            default "Hello"; //indent:12 exp:12 or 8
} //indent:0 exp:0
```

Should it be 8 or 12?

rnveach commented:
@kukreja-vikramaditya Please rebase off master as there are conflicts.

> I think the right ques to ask is what should be the indentation of L3?

Going by how methods currently work, this is the requested indentation for default module:
````
public class TestClass { // 0
    public // 4
        void // 8
        method() // 8
        throws Exception // 8 (can be adjusted by throwsIndent)
}
````

The better question is how IDEs should format them, as reminder that is what we are trying to mimic.
I am fine for keeping them all lined up like this for now.
So to answer your question, 8.

romani commented:
please rebase on latest code to resolve TC problem, TC problems are resolved in our master.

kukreja-vikramaditya commented:
@rnveach 
That is exactly the behavior I am getting. Do you want me to regenerate the reports after updating PR? Btw I based it on `MethodDefHandler` as discussed and removed only `checkModifiers()`. Please guide in case `checkType()` should be removed as well

rnveach commented:
> Do you want me to regenerate the reports after updating PR? 

Yes.

> Please guide in case checkType() should be removed as well

I will.

kukreja-vikramaditya commented:
@rnveach 
I updated the PR, but the reports did not show even a single change. Let me know if `checkType()` can be gotten rid of too

kukreja-vikramaditya commented:
TeamCity failure is related to 31 > 30 executable statements in HandlerFactory. I however have updated suppression file to exclude HandlerFactory. My guess is it isn't using my updated supressions for testing purposes?

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/ClassDefHandler.java:
```diff
@@ -68,19 +68,24 @@ protected DetailAST getListChild() {
         final DetailAST modifiers = getMainAst().findFirstToken(TokenTypes.MODIFIERS);
+        final DetailAST ident = getMainAst().findFirstToken(TokenTypes.IDENT);
+        if (getMainAst().getType() == TokenTypes.ANNOTATION_DEF) {
+            checkWrappingIndentation(getMainAst(), modifiers);
```
I removed this and got no differences in test. Please explain why it is needed and have a example in input to show it.

kukreja-vikramaditya replied:
Done

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/ClassDefHandler.java:
```diff
@@ -97,12 +102,38 @@ public void checkIndentation() {
+    private void checkAnnotatioDefIndent(final DetailAST identAst) {
+        final DetailAST atAst = getMainAst().findFirstToken(TokenTypes.AT);
+        final String atSubType = "@";
+        if (areOnSameLine(atAst, identAst) && !getIndent().isAcceptable(getLineStart(atAst))) {
```
We should only validate tokens that start the line. You do no verification that this token, and the ones below, starts the line.
if you debug this, `isAcceptable` should never be called for non-start tokens.

Could we use `LineWrappingHandler` here with a supplied first indent to do this all for us?

kukreja-vikramaditya replied:
Done. Shifted to line wrapping check.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/ClassDefHandler.java:
```diff
@@ -97,12 +102,38 @@ public void checkIndentation() {
      * @return handler name for this class.
      */
-    private static String getHandlerName(DetailAST ast) {
+    private static String getHandlerName(final DetailAST ast) {
```
undo superfluous change.

kukreja-vikramaditya replied:
removed.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/AnnotationFieldDefHandler.java:
```diff
@@ -0,0 +1,100 @@
+        super(indentCheck, ast, parent);
+    }
+
+    @Override
```
@kukreja-vikramaditya I removed all methods from this class except the constructor.
The only differences in test I saw was message changing from `annotation field def type` to `annotation field def modifier`. There was no differences in requested indentations or missing violations.

Please explain why we need all this code if existing code handles everything fine outside of small name difference.

kukreja-vikramaditya replied:
@rnveach 
Wow! Yes you are right.. i tried the same and except for the subtype name there is absolutely no difference. Would it be best to remove this handler and just work with MethodDefHandler?

I have one more example I need your help in,

```java
@ //indent:0 exp:0
 interface //indent:1 exp:0 warn
MyAnnotationWrapped { //indent:0 exp:0
    int value() default -1; //indent:4 exp:4
    AnnotationWithTarget //indent:4 exp:4
        .AnnotationInnerLineWrap field(); //indent:8 exp:4 or 8? //Error line
} //indent:0 exp:0
```

When I tried that in intellij, it indented it at 12 (8 from ident above) whereas CS expects it to be 4 (0 from ident) (lineWrappingIndentation = 4)

kukreja-vikramaditya replied:
A previously used example may be related ?
```java
    public //indent:4 exp:4
        String myString() //indent:8 exp:8
        default "Hello"; //indent:8 exp:8
```

to the above mentioned example?
Maybe this is expected behaviour for MethodDefHandler cases?

kukreja-vikramaditya replied:
Also for the 1st example, where it is assumed I am working with only MethodDefHandler,

take these 2 cases - 
Case 1 :
```java
@ //indent:0 exp:0
 interface //indent:1 exp:0 warn
MyAnnotationWrapped { //indent:0 exp:0
    int value() default -1; //indent:4 exp:4
    AnnotationWithTarget //indent:4 exp:4
        .AnnotationInnerLineWrap field(); //indent:8 exp:8 //LINE IN QUESTION
} //indent:0 exp:0
```
When CS runs on this, i get error message that it `exp` should be 4.


2. Case 2 : I try to do as suggested
```java
@ //indent:0 exp:0
 interface //indent:1 exp:0 warn
MyAnnotationWrapped { //indent:0 exp:0
    int value() default -1; //indent:4 exp:4
    AnnotationWithTarget //indent:4 exp:4
    .AnnotationInnerLineWrap field(); //indent:4 exp:4 //LINE IN QUESTION
} //indent:0 exp:0
```
Here I get the error that `exp` should be 8

rnveach replied:
> Would it be best to remove this handler and just work with MethodDefHandler?

This is only dependent on if you can prove a worth for the new class. If current code does violations and requested indentation fine, then I see no reason to keep the class unless you can give me one.
If message naming convention is only needed change and it makes the specific token easier to identify, then let's find a way to do it in `MethodDefHandler`.

>         .AnnotationInnerLineWrap field(); //indent:8 exp:4 or 8? //Error line

8, it is a wrapped line. IntelliJ probably uses 8 for linewrap indentation, which is why it wants that position.

> Maybe this is expected behaviour for MethodDefHandler cases?

For now it is expected.

rnveach commented:
> My guess is it isn't using my updated supressions for testing purposes?

Suppression done was for checkstyle checks. TC uses another system and you must suppress them another way.
I believe TC uses `@noinspection RuleName` in the javadoc.

kukreja-vikramaditya commented:
@rnveach 
I completely removed my `AnnotationFieldDefHandler` and updated `MethodDefHandler`. Reports were generated again..It was same as earlier reports except error message in a few cases. I have added a method to check indentation of `@`, `interface`, ident and `lCurly`. Should these be replaced with line wrapping checking with 0 wrapping indent?

rnveach commented:
@kukreja-vikramaditya Please reply points that are Done.

>  Should these be replaced with line wrapping checking

This was my point to you in the previous review at https://github.com/checkstyle/checkstyle/pull/4217#discussion_r113047066 . If `LineWrappingHandler` can do what we want here, then we won't need new code to do it.
You need to convince me that your code is the only way to do things right and that we need it.

kukreja-vikramaditya commented:
@rnveach 
I have made the required changes. On generating the reports, I got the same warnings except 2 which were not shown anymore, 
https://kukreja-vikramaditya.github.io/checkstyle-diff-reports/fix-issue-3134/diff/hibernate-orm/xref/home/vikramaditya/Desktop/CS/contribution-master/checkstyle-tester/repositories/hibernate-orm/hibernate-core/src/main/java/org/hibernate/annotations/FetchProfile.java.html#L49
and the other in all 3 jdks https://kukreja-vikramaditya.github.io/checkstyle-diff-reports/fix-issue-3134/diff/openjdk8/xref/home/vikramaditya/Desktop/CS/contribution-master/checkstyle-tester/repositories/openjdk8/test/java/lang/reflect/Constructor/TestParameterAnnotations.java.html#L99

New code is not finding these warnings..

Also, I am facing issue in using the 
```java
          checkWrappingIndentation(getMainAst(), getListChild(), 0,
                        getIndent().getFirstIndentLevel(), false);
```
where getMainAst -> ANNOTATION_DEF
getListChild -> OBJBLOCK of ANNOTATION DEF ( `{` )
0 -> zero wrapping as these are valid
```java
@
interface
IdentAnnotation
{
```
(as copied from Intellij's behaviour)
.
The issue is that already modifiers are being checked therefore there is a repetition for indentation check when I use `getMainAst()` (atleast that is what I think). So what I tried was to use `AT` ast but tests failed due to NullPointerException(most of them). Could you please tell me what should be parameters for checkWrappingIndentation (generally? as I earlier faced this in my other PR as well) when we want to use custom wrapping levels.

I also tried looking for any existing examples in the project but couldn't find any

romani commented:
@rnveach , what is a status of this PR ?
We do not need hanging PRs if they are in bad and unlikely to be successful.

kukreja-vikramaditya commented:
@romani 
I can complete the PR. 


@rnveach 
Can you advice on my previous comment?

rnveach commented:
Please rebase to fix conflicts and failing CI.

> So what I tried was to use AT ast but tests failed due to NullPointerException(most of them)
> ). Could you please tell me what should be parameters for checkWrappingIndentation 

I don't see anything wrong with how your trying to use it.
NPE is because `LineWrappingHandler` expected first node to always have children. `AT` node has no children.
I tried fixing the NPE by using the current node instead of grabbing the first child, and I got no regression issues in tests. I am testing out full projects regression to see if anything shows up. I'll know in the morning.
So I don't see anything else wrong with that if it makes using the line wrapping work as we want.

kukreja-vikramaditya commented:
@rnveach 
I rebased off of latest master..
Travis CI failure is unclear (network issue?) , TeamCity however is saying suppression is unnecessary in HandlerFactory.. should I remove `@noinspection`

rnveach commented:
@kukreja-vikramaditya 
>  The forked VM terminated without properly saying goodbye.

Travis issue is known problem, ignore it.

> TeamCity however is saying suppression is unnecessary ... should I remove @noinspection

CI must pass, so yes, it must be removed.

kukreja-vikramaditya commented:
@rnveach Please do review. All CIs passed

kukreja-vikramaditya commented:
@romani 
Please do review. Have rebased onto latest master
